src/http/api/utils/add-experimental.js

```diff
@@ -0,0 +1,146 @@
+'use strict'
+
+const fs = require('fs')
+const path = require('path')
+const { EOL } = require('os')
+const { Readable } = require('stream')
+const glob = require('fast-glob')
+const StreamConcat = require('stream-concat')
+const del = require('del')
+const content = require('content')
+const { Parser } = require('ipfs-multipart')
+
+const processAndAdd = (uuid, filesDir, request, reply) => {
+  // all chunks have been received
+  // TODO : here we have full size we can calculate the number of chunks to validate we have all the bytes
+  const base = path.join(filesDir, uuid) + '-'
+  const pattern = base + '*'
+  const files = glob.sync(pattern)
+
+  files.sort((a, b) => {
+    return Number(a.replace(base, '')) - Number(b.replace(base, ''))
+  })
+
+  let fileIndex = 0
+  const nextStream = () => fileIndex === files.length
+    ? null
+    : fs.createReadStream(files[fileIndex++])
+
+  createMultipartReply(
+    new StreamConcat(nextStream),
+    request,
+    reply,
+    (err) => {
+      if (err) {
+        return reply(err)
+      }
+      del(pattern, { force: true })
+        .then(paths => {
+          console.log('Deleted files and folders:\n', paths.join('\n'))
+        })
+        .catch(console.error)
+    }
+  )
+}
+
+const matchMultipartEnd = (file, boundary, cb) => {
+  const buffer = Buffer.alloc(56)
+  const fs = require('fs')
+  fs.open(file, 'r', (err, fd) => {
+    if (err) {
+      cb(err)
+    }
+
+    fs.fstat(fd, (err, stats) => {
+      if (err) {
+        cb(err)
+      }
+
+      fs.read(fd, buffer, 0, buffer.length, stats.size - 58, function (e, l, b) {
+        cb(null, b.toString().includes(boundary))
+      })
+      fs.close(fd)
+    })
+  })
+}
+
+const parseChunkedInput = (request) => {
+  const input = request.headers['x-chunked-input']
+  const regex = /^uuid="([^"]+)";\s*index=(\d*)/i
+
+  if (!input) {
+    return null
+  }
+  const match = input.match(regex)
+
+  return [match[1], Number(match[2])]
+}
+
+const createMultipartReply = (readStream, request, reply, cb) => {
+  const boundary = content.type(request.headers['content-type']).boundary
+  const ipfs = request.server.app.ipfs
+  const query = request.query
+  const parser = new Parser({ boundary: boundary })
+  const replyStream = new Readable({ read: () => {} })
+  const serialize = d => JSON.stringify(d) + EOL
+  const progressHandler = (bytes) => {
+    replyStream.push(serialize({ Bytes: bytes }))
+  }
+  // ipfs add options
+  const options = {
+    cidVersion: query['cid-version'],
+    rawLeaves: query['raw-leaves'],
+    progress: query.progress ? progressHandler : null,
+    onlyHash: query['only-hash'],
+    hashAlg: query.hash,
+    wrapWithDirectory: query['wrap-with-directory'],
+    pin: query.pin,
+    chunker: query.chunker
+  }
```

Comment:
We should extract this logic into a function e.g. `queryToAddOptions` in `http/api/resouces/files.js` and use it in `add` and `addExperimental` and just pass the result of calling it to this function. It'll save us having to keep track of another place where these options are constructed.